_posts/2015/2015-06-29-git-migrate.markdown

```diff
@@ -0,0 +1,152 @@
+
+So what happens when you run into the situation I just described? Are you stuck? Heavens no! [The thing about Git](http://2ndscale.com/rtomayko/2008/the-thing-about-git) is that its very design supports fixing up mistakes after the fact. It's very forgiving in this regard. For example, a recent blog post on the GitHub blog highlights all the different ways [you can undo mistakes in Git](https://github.com/blog/2019-how-to-undo-almost-anything-with-git).
+
+### The Easy Case - Fixing `master`
+
+This is the simple case. I made commits on master that were intended for a branch off of master. Lets walk through this scenario step by step with some visual aids.
+
+The following diagram shows the state of my repository before I got all itchy trigger finger on it.
+
+![Initial state](https://cloud.githubusercontent.com/assets/19977/8390497/bbde5f8e-1c4c-11e5-9760-9e94236423d6.png)
+
+As you can see, I have two commits to the `master` branch. HEAD points to the tip of my current branch. You can also see a remote tracking branch named `origin/master` (_this is a special branch that tracks the `master` branch on the remote server_). So at this point, my local `master` matches the `master` on the server.
+
+This is the state of my repository when I am struck by inspiration and I start to code.
+
+![First](https://cloud.githubusercontent.com/assets/19977/8390499/bbe3174a-1c4c-11e5-87ee-75aacf1a197a.png)
+
+I make one commit. Then two.
+
+![Second Commit - fixing time](https://cloud.githubusercontent.com/assets/19977/8390498/bbe2c524-1c4c-11e5-98c3-7526ae2277f9.png)
+
+Each time I make a commit, the local `master` branch is updated to the new commit. Uh oh! As in the scenario in the opening paragraph, I meant to create these two commits on a new branch creatively named `new-branch`. I better fix this up.
+
+The first step is to create the new branch. We can create it and check it out all in one step.
+
+`git checkout -b new-branch`
+
+![checkout a new branch](https://cloud.githubusercontent.com/assets/19977/8411869/0ba24220-1e3b-11e5-9895-42c486709937.png)
+
+At this point, both the `new-branch` and `master` point to the same commit. Now I can force the master branch back to its original position.
+
+`git branch --force master origin/master`
+
+![force branch master](https://cloud.githubusercontent.com/assets/19977/8411975/b05c941e-1e3b-11e5-8e84-f36535fb7893.png)
+
+Here's the set of commands that I ran all together.
+
+```bash
+git checkout -b new-branch
+git branch --force master origin/master
+```
+
+### Fixing up a non-master branch
+
+![The wrong branch](https://cloud.githubusercontent.com/assets/19977/8369613/48019364-1b71-11e5-9a28-b748a2802ed7.png)
+
+This case is a bit more complicated. Here I have a branch named `wrong-branch` that is my current branch. But I thought I was working in the `master` branch. I make two commits in this branch by `mistake` which causes this fine mess.
```

Comment:
Why is "mistake" highlighted like code? Maybe you have a program called `mistake` that you can blame when things go wrong. (That could come in handy...)